src/osx/core/uilocale.mm

```diff
@@ -0,0 +1,187 @@
+///////////////////////////////////////////////////////////////////////////////
+// Name:        src/osx/core/uilocale.mm
+// Purpose:     wxUILocale implementation for macOS
+// Author:      Vadim Zeitlin
+// Created:     2021-08-01
+// Copyright:   (c) 2021 Vadim Zeitlin <vadim@wxwidgets.org>
+// Licence:     wxWindows licence
+///////////////////////////////////////////////////////////////////////////////
+
+// ============================================================================
+// declarations
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// headers
+// ----------------------------------------------------------------------------
+
+// for compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+#if wxUSE_INTL
+
+#include "wx/uilocale.h"
+#include "wx/private/uilocale.h"
+
+#include "wx/osx/core/cfref.h"
+#include "wx/osx/core/cfstring.h"
+
+#include <Foundation/NSString.h>
+#include <Foundation/NSLocale.h>
+
+extern wxString
+wxGetInfoFromCFLocale(CFLocaleRef cfloc, wxLocaleInfo index, wxLocaleCategory cat);
+
+// ----------------------------------------------------------------------------
+// wxLocaleIdent::GetName() implementation using Foundation
+// ----------------------------------------------------------------------------
+
+wxString wxLocaleIdent::GetName() const
+{
+    // Construct name in right format:
+    // MacOS: <language>-<script>_<REGION>
+
+    wxString name;
+    if (!m_language.empty())
+    {
+        name << m_language;
+
+        if (!m_script.empty())
+        {
+            name << "-" << m_script;
+        }
+
+        if (!m_region.empty())
+        {
+            name << "_" << m_region;
+        }
+    }
+
+    return name;
+}
+
+namespace
+{
+
+// ----------------------------------------------------------------------------
+// wxUILocale implementation using Core Foundation
+// ----------------------------------------------------------------------------
+
+class wxUILocaleImplCF : public wxUILocaleImpl
+{
+public:
+    // Takes ownership of NSLocale.
```

Comment:
It's a private/internal class so I thought it would be ok to take ownership here, even if it's going to make the code slightly more complicated (because we'll need an extra `release` in `Create()`).

I was really surprised, but apparently there is no smart pointer for Objective C++ code in wx similar to `wxCFRef` -- or did I miss it?

Reply:
> It's a private/internal class so I thought it would be ok to take ownership here, even if it's going to make the code slightly more complicated (because we'll need an extra `release` in `Create()`).

the call you use currently there `[NSLocale localeWithLocaleIdentifier:...]` is a class factory method, you are just supposed to use the object instance with these kind of calls within the method, not release if after https://developer.apple.com/library/archive/documentation/General/Conceptual/CocoaEncyclopedia/ClassFactoryMethods/ClassFactoryMethods.html  . So if you intend to keep it longer, you have to retain it.

> I was really surprised, but apparently there is no smart pointer for Objective C++ code in wx similar to `wxCFRef` -- or did I miss it?

there is `template <class T> class wxNSObjRef`, since CFLocale and NSLocale are toll-free bridged you could also use a wxCFRef. When I have a obj-c++ class, then I'm using the native properties for these attributes.

Reply:
> I did use `wxCFRef` originally, but I didn't want to have all the casts, so I got rid of it. But I've somehow completely missed `wxNSObjRef`, thanks. However it also takes ownership of the pointer in its ctor, so it's not really going to help here, is it?

right, also when using `wxNSObjRef`, we will have to `retain` explicitly, because these are not `create` Methods, however we could get rid of the `release` in our class.

> And I'm not sure I understand properties well enough to be sure I use them correctly. Should I just have `@property (retain)` here?

with pure c++ classes I don't think you can use properties.

> BTW, I build with ASAN (`-fsanitize=address`) under Mac and yet it doesn't detect memory leaks somehow, I wonder what am I doing wrong -- or could it really just not work correctly under ARM?

I don't really know. You'd need a autorelease pool in place, when running these methods to detect it at all, so that destroying this memory pool would lead to a double delete detection.

Reply:
OK, for now I'll stay with the manual retain/release then.